libafl_frida/src/asan/asan_rt.rs

```diff
@@ -526,32 +537,85 @@ impl AsanRuntime {
                 paste::paste! {
                     log::trace!("Hooking {}", stringify!($name));
                     let target_function = frida_gum::Module::find_export_by_name($lib, stringify!($name)).expect("Failed to find function");
-                    self.hooks.insert(stringify!($name).to_string(), target_function);
+
+                    static mut [<$name:snake:upper _PTR>]: *const c_void = 0 as *const c_void;
+                    unsafe {[<$name:snake:upper _PTR>] = target_function.0}
+                    //extern "system" { fn $name($($param: $param_type),*) -> $return_type; }
+
 
                     #[allow(non_snake_case)]
                     unsafe extern "C" fn [<replacement_ $name>]($($param: $param_type),*) -> $return_type {
                         let mut invocation = Interceptor::current_invocation();
                         let this = &mut *(invocation.replacement_data().unwrap().0 as *mut AsanRuntime);
-                        if this.hooks_enabled && this.[<hook_check_ $name>]($($param),*) {
+                        if this.hooks_enabled && this.[<hook_check_ $name>]($($param),*){
                             let previous_hook_state = this.hooks_enabled;
                             this.hooks_enabled = false;
                             let ret = this.[<hook_ $name>]($($param),*);
                             this.hooks_enabled = previous_hook_state;
                             ret
                         } else {
-                            let original = std::mem::transmute::<*const c_void, extern "C" fn($($param: $param_type),*) -> $return_type>(this.hooks.get(&stringify!($name).to_string()).unwrap().0);
+                            let original = unsafe {std::mem::transmute::<*const c_void, extern "C" fn($($param: $param_type),*) -> $return_type>([<$name:snake:upper _PTR>])};
                             let previous_hook_state = this.hooks_enabled;
                             this.hooks_enabled = false;
                             let ret = (original)($($param),*);
                             this.hooks_enabled = previous_hook_state;
                             ret
                         }
+
                     }
+
+                    let self_ptr = core::ptr::from_ref(self) as usize;
                     let _ = interceptor.replace(
                         target_function,
                         NativePointer([<replacement_ $name>] as *mut c_void),
-                        NativePointer(core::ptr::from_mut(self) as *mut c_void)
+                        NativePointer(self_ptr as *mut c_void)
                     );
+                    self.hooks.push(target_function);
+                }
+            }
+        }
+        //This is used for cases where the hook needs a pointer to the original function
+        #[cfg(windows)]
+        macro_rules! hook_func_nolinkage {
```

Comment:
I think the macros name is misleading. Also: Let's just go ahead and make it the default to *always* add the original function pointer to the hook function signature, but let's make it the first parameter (after `&mut self`). So you can merge this macro into the two above.